internal/resolution/variablesources/olm_variable_source.go

```diff
@@ -0,0 +1,81 @@
+package variablesources
+
+import (
+	"context"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/operator-framework/deppy/pkg/deppy"
+	rukpakv1alpha1 "github.com/operator-framework/rukpak/api/v1alpha1"
+
+	operatorsv1alpha1 "github.com/operator-framework/operator-controller/api/v1alpha1"
+	olmvariables "github.com/operator-framework/operator-controller/internal/resolution/variables"
+)
+
+type OLMVariableSource struct {
+	client        client.Client
+	catalogClient BundleProvider
+}
+
+func NewOLMVariableSource(cl client.Client, catalogClient BundleProvider) *OLMVariableSource {
+	return &OLMVariableSource{
+		client:        cl,
+		catalogClient: catalogClient,
+	}
+}
+
+func (o *OLMVariableSource) GetVariables(ctx context.Context) ([]deppy.Variable, error) {
+	operatorList := operatorsv1alpha1.OperatorList{}
+	if err := o.client.List(ctx, &operatorList); err != nil {
+		return nil, err
+	}
+
+	bundleDeploymentList := rukpakv1alpha1.BundleDeploymentList{}
+	if err := o.client.List(ctx, &bundleDeploymentList); err != nil {
+		return nil, err
+	}
+
+	allBundles, err := o.catalogClient.Bundles(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	requiredPackages := []*olmvariables.RequiredPackageVariable{}
+	for _, operator := range operatorList.Items {
+		requiredPackage, err := RequiredPackageVariable(allBundles, operator.Spec.PackageName, operator.Spec.Channel, operator.Spec.Version)
+		if err != nil {
+			return nil, err
+		}
+		requiredPackages = append(requiredPackages, requiredPackage)
+	}
+
+	installedPackages, err := InstalledPackageVariables(allBundles, bundleDeploymentList.Items)
+	if err != nil {
+		return nil, err
+	}
+
+	bundles, err := BundleVariables(allBundles, requiredPackages, installedPackages)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
It feels off to me that we're using variables to compute more variables. For me at least, this structure makes it really hard to build intuition about what each variable is actually adding to the problem.

Without diving too deeply, it appears to me like we:
- Figure out what bundles are required based on the `Operator` objects that exist (`requiredPackages`)
- Figure out what bundles are required based on existing BDs and their successors (but this is for all BundleDeployments, not just BDs managed by Operators.  (`installedPackages`)
- For `requiredPackage` bundles and `installedPackage` bundles combined, build a set of `Bundle` variables that:
   1. include all `requiredPackage` and `installedPakcage` bundles (so `bundles` is a superset?)
   2. include all possible dependencies, recursively, based on the roots from (i)


Am I getting this correct? If so, what purpose do the `requiredPackages` and `installedPackages` variables actually serve for the resolver? If none (because `bundles` is a superset?), can we refactor such that `requiredPackages` and `installedPackages` are not actually variables and are instead just lists that we use to filter `allBundles` to `bundles`?

Reply:
> However on failed resolution error messages will be harder to understand: there will be no difference between a dependency of a dependency of a dependecny and a thing which user explicitly requested.

I'm not convinced that's true (or at least necessarily has to be the case). I'd imagine something like this is what is (or could be) fed into the resolver:
- OperatorVariable(Package: a, DependsOnOneOf(aCurrent, aSuccessor), Mandatory: true)
- BundleVariable(Bundle: aCurrent, DependsOnOneOf(bCandidate), Mandatory: false)
- BundleVariable(Bundle: aSuccessor, DependsOnOneOf(bCandidate), Mandatory: false)
- BundleVariable(Bundle: bCandidate, DependsOnOneOf(cCandidate), Mandatory: false)
- BundleVariable(Bundle: cCandidate, Mandatory: false)

Any resolution failure that involves `cCandidate` being in conflict will only be unsat in this case because of the mandatory Operator variable for package `a`. And the unsat error message would traverse this dependency chain, right?

Reply:
> Both `RequiredPackageVariable` and `InstalledPackageVariable` represent bundles so is `BundleVariable`

Just realised - this is incorrect. `RequiredPackageVariable` and `InstalledPackageVariable` represent packages which depend on specific bundles.

> I'd imagine something like this is what is (or could be) fed into the resolver

Do I understand correctly that it will be an equivalent of combining `RequiredPackageVariable` and `InstalledPackageVariable` together into `OperatorVariable`? At first glance it makes sense and might work, but we need to look into it in more depth.

I suggest that we do it as a separate effort: in this PR leave the structure of the resolution problem as it currently is in `main` (this PR is already large) and separately look into combining variables.

I'll split #437 into two issues: geting rid of variable sources and combining variables.